Initial implementation

wojtekmach opened the pull request:
This is the implementation of the [OTP macOS builds](https://github.com/erlef/build-and-packaging-wg/issues/80) proposal.

Updated README.md: https://github.com/erlef/otp_builds/blob/wm-initial/README.md

As you can see in commits list, the GitHub job automatically commits `builds/*.txt` changes.

Builds can be manually triggered like this:

```
ref=OTP-27.0.1;
gh workflow run build.yaml \
  -R erlef/otp_builds \
  --ref wm-initial \
  -f otp-ref-name=$ref \
  -f otp-ref=$(gh api repos/erlang/otp/commits/$ref --jq .sha)
```

paulo-ferraz-oliveira commented on .github/workflows/build.yaml:
```diff
@@ -1,8 +1,98 @@
+  schedule:
+    name: "Schedule nightly"
+    if: github.event_name == 'schedule'
     runs-on: ubuntu-latest
```
Would it make sense to version this? I've had cases where CI broke for similar reasons (using latest). Or maybe we want it and we'll deal with issues once they happen (?)

paulo-ferraz-oliveira replied:
Was just a detail, in any case, can be dealt with later. What'll probably be important is to have a "Job failed" issue, but I already created an issue to deal with that later.

paulo-ferraz-oliveira commented on .github/workflows/build.yaml:
```diff
@@ -1,8 +1,98 @@
+      OTP_DIR: /tmp/builds/otp-${{ inputs.otp-ref-name }}-${{ inputs.otp-ref }}-openssl-${{ inputs.openssl-version }}-wxwidgets-${{ inputs.wxwidgets-version }}-macos-${{ matrix.pair.arch }}
+      OTP_REF: ${{ inputs.otp-ref }}
+    strategy:
+      matrix:
```
Would it make sense to add:

* `fail-fast: true` - e.g. failing 13 shouldn't prevent 14 from getting released, or vice-versa
* `max-parallel: 1` - this way we're bound to have less concurrency issues with GitHub runners (I've had instances where the filesystem seems shared - or some kind of caching is taking place - that this option eased)

wojtekmach replied:
I added fail-fast: false for now, thanks!

paulo-ferraz-oliveira commented on scripts/build_otp_macos.bash:
```diff
@@ -0,0 +1,144 @@
+  fi
+}
+
+build_otp() {
```
Should we also make sure that static linking was done properly?

Something like

```
    crypto=$(find . -name "crypto.so")
    dyn=$(otool -L "${crypto}")
    dyn=$(echo "${dyn}" | tail -n +2)
    if [[ ${dyn} -ne "" ]]; then
        echo "OpenSSL linking not static! Exiting..."
        exit 1
    fi
```

or similar (?)

starbelly replied:
Agreed, this is a solid sanity check. Alternatively, one can use `dyld_info crypto.so` | grep -q openssl`, if the exit status is 0, then we know static linking failed.

wojtekmach replied:
Good call, done in c87f860.

paulo-ferraz-oliveira commented on scripts/build_otp_macos.bash:
```diff
@@ -0,0 +1,144 @@
+    export ERL_TOP=$PWD
+    export ERLC_USE_SERVER=true
+
+    ./otp_build configure \
+      --with-ssl=${OPENSSL_DIR} \
+      --disable-dynamic-ssl-lib
```
Where to we pick the version to build? Do we rely only on the `workflow_dispatch`? (sorry, don't know what the current strategy is) Or would it make sense to have a "minimum supported version" and then try to figure out the ones that are missing from that reference compared to `builds/*.txt`?

paulo-ferraz-oliveira replied:
(we can leave this for a further issue/implementation if the idea is sound)

wojtekmach replied:
Yes, it's workflow_dispatch input. The script can be executed locally too, the usage is `build_otp_macos.bash ref_name`.

paulo-ferraz-oliveira commented on .github/workflows/build.yaml:
```diff
@@ -1,8 +1,98 @@
+        run: |
+          tgz="$OTP_REF_NAME-macos-${{ matrix.pair.arch }}.tar.gz"
+          tar czf $tgz --cd $OTP_DIR .
+          gh release create  $OTP_REF_NAME || true
```
Maybe we could add notes, e.g. like `This is Erlang/OTP <vsn> compiled for Darwin <vsn> (64 bit).`

wojtekmach replied:
I added these notes for now:

```
Automated build for https://github.com/erlang/otp/tree/{ref}
```

paulo-ferraz-oliveira commented on README.md:
```diff
@@ -1 +1,49 @@
+
+Supported architectures:
+
+  * `amd64`
+  * `arm64`
```
Should this state that AMD is only for 13, and ARM is only for 14 (current GitHub reality?)

paulo-ferraz-oliveira replied:
Might not be important, then. I was just stating that the runner for 13 is AMD (from GitHub) and the one for 14 is ARM.

paulo-ferraz-oliveira commented on README.md:
```diff
@@ -1 +1,49 @@
+
+## List of Builds
+
+List of builds can be found here:
```
Should this mention that `master` and `maint*` are also available but they "move" over time?

wojtekmach replied:
Yup, done!

paulo-ferraz-oliveira commented on README.md:
Should this mention that documentation is available as per Erlang/OTP's defaults, `chunks`?

paulo-ferraz-oliveira replied:
Is it, though? I ran into an issue, elsewhere, that e.g. for 13 + 27.0-rc2 `ex_doc` failed to download because of a missing `sha256sum` (it's since been fixed to include `shasum` in the verification). I'm not sure we want to have docs with the builds and/or we should have a mechanism to bypass that when doc build fails.

wojtekmach replied:
Sorry, I don't follow. Could you explain in more detail what you tried and what didn't work?

paulo-ferraz-oliveira replied:
To generate the docs for `27.0-rc2`, for example, you need `ex_doc`, but at that version Erlang/OTP's compilation failed on missing `sha256sum` (IIRC) on the host. These days it accepts `shasum` ("native" to macOS). But since `27.0` didn't require it we may want to skip that lest we create an exception only for 1-2 `rc`s.

wojtekmach replied:
I don't think we need to generate html docs here, people can grab them from erlang.org. I'd focus on just doc _chunks_ here.

paulo-ferraz-oliveira replied:
Got it. Then maybe the README should mention it, if found pertinent. Otherwise feel free to resolve the conversation 👍

paulo-ferraz-oliveira commented on README.md:
For clarity, should we mention how OpenSSL and wxWidgets are build and/or statically linked?

And what about "this is our strategy for updating OpenSSL if a vulnerability is found" (?)

paulo-ferraz-oliveira commented on README.md:
```diff
@@ -1 +1,49 @@
+to the `{ref_name}`, `{datetime}` is the time the build was created and `{sha256}` is the build
+SHA-256 checksum.
+
+Example `builds/macos-arm64.txt` entry:
```
Should we mention something similar to 

"
At a very minimum, when consuming the images generated by this repository, we strongly suggest you
verify their SHA sum the same way we generate it ("<method>" found next to `.tar.gz`)
"

starbelly replied:
^ that. I would suggest opening PRs @paulo-ferraz-oliveira , we need to strong signaling here in regard to security.

paulo-ferraz-oliveira replied:
Yeah, I'll wait for @wojtekmach's comments. Whatever he doesn't think is relevant for this PR we can turn into issues and implement later.

wojtekmach replied:
Thanks, I added the following to the README:

```
After downloading the build you should verify its integrity against builds csv mentioned in the previous section, for example:

    $ curl -fLO https://github.com/erlef/otp_builds/releases/download/OTP-27.1.2/OTP-27.1.2-aarch64-apple-darwin.tar.gz
    $ checksum=`curl -fsSL https://github.com/erlef/otp_builds/raw/builds/aarch64-apple-darwin.csv | grep OTP-27.1.2, | cut -d"," -f4`
    $ sha256 OTP-27.1.2-aarch64-apple-darwin.tar.gz --check $checksum
```

paulo-ferraz-oliveira reviewed (approved):
Left a few comments but nothing that can't be turned into GitHub issues and done later. Approving...

tsloughter commented on scripts/build_otp_macos.bash:
```diff
@@ -0,0 +1,144 @@
+  local url="https://github.com/openssl/openssl"
+
+  if [ ! -d ${src_dir} ]; then
+    git clone --depth 1 ${url} --branch ${ref_name} ${src_dir}
```
Should this be done in the action so it doesn't have to be done for build macos, windows, ubuntu, etc separately? Could do it once and use `git worktree` to checkout a clean copy to a different directory for each one.

Also had the random thought that maybe this reop should contain a fork so that git tags/releases pointed to something real. The `main` branch would still be as it is but have like `upstream-master` and `upstream-maint` then the nightly script updates those by pulling and pushing from upstream, which causes the build actions to kickoff.

wojtekmach replied:
I'd prefer to keep as much as possible in the sh script so it can be tested locally. I'd extract stuff out when we add ubuntu. Perhaps instead of build_otp_macos.sh we have build_otp.sh, i.e. it's similar enough across the board that it's easier to have one script than a bunch. WDTT?

wojtekmach replied:
Regarding making this a fork, brilliant, I think this works. I'll think about it some more but I can't think of a downside, I really like the idea that releases would point to something real (even if forked) and we'd keep it totally unchanged and SHAs will match with upstream.

master/upstream-master is a weird edge case. We need a tag for it so we could have a release so we could store assets for it. The release points to a tag and a SHA and I determined that we can't change the SHA on the release. However (kinda doh) we can remove the tag and create a new one pointing to updated SHA and force push the tag. It seems to work!

WDYT about calling it `master-nightly` and `maint-nightly` instead?

tsloughter replied:
Ah yea, `-nightly` works nicely.

paulo-ferraz-oliveira replied:
> maybe this reop should contain a fork

Wouldn't a branch suffice, in this case? As long as you keep the name known. You'd benefit from keeping the logic in the same repo.

I'm not sure I understood the last comment by @wojtekmach, though. Do you prefer a fork or a branch? And what would you see as a downside to having a branch?

wojtekmach replied:
I'd have to double check it but I vaguely remember that when a repo is a _fork_ GitHub adds some special handling to it, perhaps if you push a commit that @-mentions someone it does _not_ fire a notification from the fork. If there's nothing like that then yeah, doesn't matter if it's a fork or not and I'd keep it as is. But yes we are definitely having all GitHub actions and scripts in this repo on the main branch regardless if it's a fork or not.

wojtekmach replied:
I've run into issues pushing erlang/otp tags into this repo (they contain .github/workflows changes which need special permissions). I think this is a good idea but I'd like to take it off the table for now, if someone figures it out perhaps we start doing it for OTP 28 or something. In the meantime I'm addicting release notes to our faux releases that point to the upstream ones, see e.g.: https://github.com/erlef/otp_builds/releases/tag/OTP-26.2.5

wojtekmach replied:
If we figure this out down the road I think we could simply overwrite tags, e.g. `OTP-27.1` instead of pointing to https://github.com/erlef/otp_builds/commit/b5893a3c3a8d0ab54be5d04de450b24d9e5aa149 like all our releases will point to https://github.com/erlang/otp/commit/9ae2ef54fc6dacb5bfbe2dc6fc6f2522d57ce1fb.

I'll resolve this conversation for now.

paulo-ferraz-oliveira commented on .github/workflows/build.yaml:
```diff
@@ -84,12 +84,41 @@ jobs:
       - name: "Upload"
         env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          # This is a personal access token for user wojtekmach-erlef-bot that has write access.
+          # We need a token with `workflows` permissions because by pushing git tags
+          # (from upstream, erlang/otp) we're modifying workflow files (.github/workflows)
+          # on those tags and that is restricted. Perhaps this would be not necessary if this repo
+          # was a _fork_ of erlang/otp.
```
@starbelly, it should be fairly simple to create an ErlEF bot and then share secrets with certain repositories (e.g. this one) via the Settings, so @wojtekmach doesn't need a PAT. You need to be an admin. or the org. so I guess you can do it. I can walk you through the process as I've done it before, already.

starbelly replied:
It looks like a workflow token was created, I think this is resolved now?

wojtekmach replied:
It's not working yet. I was trying to workaround permissions issues to prove the idea but no luck yet. If you could set it up I'd appreciate it.

starbelly replied:
That's odd, what you have seems correct,  that is using `GITHUB_TOKEN` and setting the permissions in the workflow file (per https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#permissions-for-the-github_token) 

If you know what the issue is, and give pass a pointer, I'm happy to resolve.

wojtekmach replied:
As I mentioned in the other comment, I'd like to postpone this indefinitely. I'm going to remove the related leftover code from the repo too.

wojtekmach commented:
@paulo-ferraz-oliveira @starbelly thank you for the feedback. Regarding https://github.com/jelly-beam/otp-macos, it is quite comprehensive and it's a great starting point. Now that you can see how I tried implementing this, we can consider the next steps. 

Would you rather:

1. merge changes from this PR to jelly-beam/otp-macos, rename it to erlef/otp-macos
2. merge changes from jelly-beam/otp-macos here
3. ignore this PR, merge jelly-beam/otp-macos as is here 

?

paulo-ferraz-oliveira commented:
> Would you rather:

"merge changes from jelly-beam/otp-macos here", for me 👍

Some stuff can be adapted, I guess, but only if there's a need for it; as long as we have the builds here there's only a few details I'd update, but I don't think they'll change the base behaviour:

* (later?) I'd start the builds from 25.1 since I think it's feasible #4 
* I'd like to have a mechanism for rebuilding a build (e.g. you update a file, then it gets picked up and replaces the build in `.txt` - or maybe the dispatch already does it?): important in case we want to update e.g. OpenSSL without disruption
* I'd prefer to have a `.sha256.txt` of the result next to it, for security #25 

This said, I'd go with your current implementation; it's a good starting point and I can see a couple of advantages:

* compiling OpenSSL (instead of `brew`ing it),
* doing all stuff by hand (instead of `kerl`ing it),
* using wxWidgets (I didn't test that as a default for `kerl` so it's possible it's not working, but it's better to have it, too)

I'm also Ok with using `otp-macos` but it'd need some issues taken care of (remove `brew` and `kerl`, for example; add wxWigets), so it's less relevant here.

starbelly commented:
> This said, I'd go with your current implementation; it's a good starting point and I can see a couple of advantages:
> 
>     * compiling OpenSSL (instead of `brew`ing it),
> 
>     * doing all stuff by hand (instead of `kerl`ing it),
> 
>     * using wxWidgets (I didn't test that as a default for `kerl` so it's possible it's not working, but it's better to have it, too)
> 
> 
> I'm also Ok with using `otp-macos` but it'd need some issues taken care of (remove `brew` and `kerl`, for example; add wxWigets), so it's less relevant here.

☝️ That. Let's go with this PR, then roll in changes that were done in otp-macos, collaboratively.

wojtekmach commented:
Sounds good, I’ll send some updates here soon!

starbelly commented:
@wojtekmach Just to confirm, you are ready for this to be merged?

wojtekmach commented:
Not quite yet, I'll merge this myself if that's OK.

wojtekmach commented on README.md:
```diff
@@ -1 +1,77 @@
+
+Example build URLs:
+
+  * <https://github.com/erlef/otp_builds/releases/download/OTP-27.0.1/OTP-27.0.1-macos-arm64.tar.gz>
```
> `OTP-27.0.1-macos-arm64.tar.gz`

I'd like to get some feedback on the naming scheme.

Ideally we'd use naming that'd be be easy to arrive at using basic tooling, like `uname`, however speaking of that it is so inconsistent:

- macOS arm64: `bash$ uname -sm #=> Darwin arm64`
- Linux arm64: `bash$ uname -sm #=> Linux aarch64`
- Windows x86_64: `bash$ uname -sm #=> MINGW64_NT-10.0 x86_64`, `powershell> echo $env:PROCESSOR_ARCHITECTURE #=> AMD64`

so I think nevermind that.

I have found these naming schemes:

1. target triple, e.g. from `rustc --print target-list`

    - `aarch64-apple-darwin`
    - `aarch64-apple-ios`
    - `aarch64-unknown-linux-gnu`
    - `aarch64-unknown-freebsd`

   (No plans for iOS or FreeBSD builds here _at the moment_.)

2. `$GOOS`/`$GOARCH`, `go tool dist list`:

    - `darwin-arm64`
    - `linux-arm64`
    - `freebsd-arm64`

As an aside, existing hexpm/bob builds, https://github.com/hexpm/bob?tab=readme-ov-file#erlang-builds use amd64/arm64 architecture identifier. I hope to eventually move them to this repor. Unless we could have Linux builds that are distribution independent (perhaps just libc dependent), we'd have to put the Ubuntu name in there _anyway_, say `OTP-27.0.1-linux-arm64-ubuntu-22.04.tar.gz` so it seems like a custom naming scheme too.

As an another aside, official OTP Windows builds are named e.g. `otp_win64_27.1.zip`.

Finally, there's a matter of `:erlang.system_info(:system_architecture)`:

  - macOS arm64 (built-on macOS 14):

        elixir -e "IO.puts :erlang.system_info(:system_architecture)"
        aarch64-apple-darwin23.5.0

  - macOS x64 (built-on macOS _13_):

        elixir -e "IO.puts :erlang.system_info(:system_architecture)"
        x86_64-apple-darwin22.6.0

  - Linux arm64:

        $ docker run --rm -it hexpm/elixir:1.17.3-erlang-27.1.2-ubuntu-jammy-20240808 elixir -e "IO.puts :erlang.system_info(:system_architecture)"
        aarch64-unknown-linux-gnu

  - Linux x64:

        $ docker run --rm -it hexpm/elixir:1.17.3-erlang-27.1.2-ubuntu-jammy-20240808 elixir -e "IO.puts :erlang.system_info(:system_architecture)"
        x86_64-pc-linux-gnu

  - Windows:

        $ elixir -e "IO.puts :erlang.system_info(:system_architecture)"
        win32


Worth reminding this is the target triple (not a triple on Windows) of the _build_, not of the currently running OS. I'm not sure if `23.5.0` in `aarch64-apple-darwin23.5.0` is useful btw. I'm making these builds with `-mmacosx-version-min=11.0`, so if anything, following `swiftc -print-target-info | jq .target.triple`, it could be `arm64-apple-macosx11.0`. `swiftc a.swift -target aarch64-apple-macos11.0` also works. I think `macos` in the target triple is nicer but `darwin` as a name is probably still more pervasive.

Not sure if we need to address both build file naming and OTP `system_architecture` but I think for these builds we could do the following which shouldn't hurt:

- `OTP-27.1.2-aarch64-apple-darwin.tar.gz` `system_architecture` reports `aarch64-apple-darwin`
- `OTP-27.1.2-x86_64-apple-darwin.tar.gz` `system_architecture` reports `x86_64-apple-darwin`

When we add Ubuntu we could do this:

- `OTP-27.1.2-aarch64-pc-linux-ubuntu-24.04.tar.gz` `system_architecture` reports `aarch64-pc-linux` (or `unknown-linux`?)
- `OTP-27.1.2-x86_64-pc-linux-ubuntu-22.04.tar.gz` `system_architecture` reports `x86_64-pc-linux`

or, ideally, Linux distribution-independent builds:

- `OTP-27.1.2-aarch64-pc-linux.tar.gz` `system_architecture` reports `aarch64-pc-linux`
- `OTP-27.1.2-x86_64-pc-linux.tar.gz` `system_architecture` reports `x86_64-pc-linux`

WDYT?

wojtekmach commented:
I have made the following changes:

1. I added a disclaimer at the top of the README saying this is a work in progress. I'd like to go over all issues and PRs and built all releases all the way back to OTP 25.0. Then I'd consider this ready.
2. I changed naming from `macos-{amd64,arm64}` to `{x86_64,aarch64}-apple-darwin` in filenames. The rationale was we'd match `erlang:system_info(system_architecture)`. I'm going to keep build artifacts using old names for a while since `beamup` was already using this project.
3. I changed `builds/*.txt` to `builds/*.csv` and added openssl and wxwidgets versions there.

I'm going to merge this as is to unblock others tasks but please feel free to keep the discussion going here or in other issues.

Thanks everyone for reviewing this!
